File: Data Structures & Algorithms/count-connected-components/submission-1.py
class DSU:
    def __init__(self, n):
        self.parents  = list(range(n))
        self.size = [1] * n

    def find(self, x):
        if self.parents[x] != x:
            self.parents[x] = self.find(self.parents[x])

        return self.parents[x]

    def union(self, u, v):
        u = self.find(u)
        v = self.find(v)

        if u == v:
            return False

        if self.size[u] < self.size[v]:
            u, v = v, u

        self.parents[v] = u
        self.size[u] += self.size[v]

        return True

    def connected(self, a, b):
        return self.find(a) == self.find(b)

class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        dsu = DSU(n)
        res = n

        for u, v in edges:
            if dsu.union(u, v):
                res -= 1
        
        return res

File: Data Structures & Algorithms/pow-x-n/submission-1.cpp
class Solution {
private:
    double helper(double x, long long n) {
        if (x == 0) {
            return 0;
        }
        if (n == 0) {
            return 1;
        }

        double res = helper(x * x, n / 2);

        if (n % 2 != 0) {
            return x * res;
        } else {
            return res;
        }
    }
public:
    double myPow(double x, int n) {        
        long long power = n;
        if (power < 0) {
            power = -power;
        }

        double res = helper(x, power);

        if (n >= 0) {
            return res;
        } else {
            return 1 / res;
        }
    }
}; 